src/Aggregate/AggregateRepository.php

```diff
@@ -0,0 +1,279 @@
+ */
+
+declare(strict_types=1);
+
+namespace Prooph\EventSourcing\Aggregate;
+
+use ArrayIterator;
+use Prooph\Common\Messaging\Message;
+use Prooph\EventSourcing\Snapshot\SnapshotStore;
+use Prooph\EventStore\EventStore;
+use Prooph\EventStore\Stream\Stream;
+use Prooph\EventStore\Stream\StreamName;
+
+/**
+ * Class AggregateRepository
+ *
+ * @package Prooph\EventSourcing\Aggregate
+ * @author Alexander Miertsch <contact@prooph.de>
+ */
+class AggregateRepository
+{
+    /**
+     * @var EventStore
+     */
+    protected $eventStore;
+
+    /**
+     * @var AggregateTranslator
+     */
+    protected $aggregateTranslator;
+
+    /**
+     * @var AggregateType
+     */
+    protected $aggregateType;
+
+    /**
+     * @var array
+     */
+    protected $identityMap = [];
+
+    /**
+     * @var SnapshotStore|null
+     */
+    protected $snapshotStore;
+
+    /**
+     * @var StreamName
+     */
+    protected $streamName;
+
+    /**
+     * @var bool
+     */
+    protected $oneStreamPerAggregate;
+
+    public function __construct(
+        EventStore $eventStore,
+        AggregateType $aggregateType,
+        AggregateTranslator $aggregateTranslator,
+        SnapshotStore $snapshotStore = null,
+        StreamName $streamName = null,
+        bool $oneStreamPerAggregate = false
+    ) {
+        $this->eventStore = $eventStore;
+        $this->eventStore->getActionEventEmitter()->attachListener('commit.pre', [$this, 'addPendingEventsToStream']);
+
+        $this->aggregateType = $aggregateType;
+        $this->aggregateTranslator = $aggregateTranslator;
+        $this->snapshotStore = $snapshotStore;
+        $this->streamName = $streamName;
+        $this->oneStreamPerAggregate = $oneStreamPerAggregate;
+    }
+
+    /**
+     * Repository acts as listener on EventStore.commit.pre events
+     * In the listener method the repository checks its identity map for pending events
+     * and appends the events to the event stream.
+     */
+    public function addPendingEventsToStream(): void
+    {
+        foreach ($this->identityMap as $aggregateId => $aggregateRoot) {
+            $pendingStreamEvents = $this->aggregateTranslator->extractPendingStreamEvents($aggregateRoot);
+
+            if (count($pendingStreamEvents)) {
+                $enrichedEvents = [];
+
+                foreach ($pendingStreamEvents as $event) {
+                    $enrichedEvents[] = $this->enrichEventMetadata($event, $aggregateId);
+                }
+
+                $streamName = $this->determineStreamName($aggregateId);
+
+                $this->eventStore->appendTo($streamName, new ArrayIterator($enrichedEvents));
+            }
+        }
+
+        //Clear identity map
+        $this->identityMap = [];
+    }
+
+    /**
+     * @param object $eventSourcedAggregateRoot
+     *
+     * @throws Exception\AggregateTypeException
+     */
+    public function addAggregateRoot($eventSourcedAggregateRoot): void
```

Comment:
@lucascourot it is not necessary to implement this interface to get working with event sourcing. That is intentional and documented.. somewhere ;)